File: app/src/main/java/jp/hazuki/yuzubrowser/bookmark/BookmarkFolder.java
package jp.hazuki.yuzubrowser.bookmark;

import android.text.TextUtils;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import jp.hazuki.yuzubrowser.bookmark.util.BookmarkIdGenerator;
import jp.hazuki.yuzubrowser.tab.manager.MainTabData;

public class BookmarkFolder extends BookmarkItem implements Serializable {
    protected static final String COLUMN_NAME_LIST = "2";
    public static final int BOOKMARK_ITEM_ID = 1;

    public BookmarkFolder parent;
    final ArrayList<BookmarkItem> list;

    public BookmarkFolder(String title, BookmarkFolder parent, long id) {
        super(title, id);
        this.list = new ArrayList<>();
        this.parent = parent;
    }

    public BookmarkFolder(List<MainTabData> list, long id) {
        super(null, id);
        this.list = new ArrayList<>(list.size());
        for (MainTabData tab : list)
            if (!TextUtils.isEmpty(tab.getUrl()))
                this.list.add(new BookmarkSite((tab.getTitle() != null) ? tab.getTitle() : tab.getUrl(), tab.getUrl(), BookmarkIdGenerator.getNewId()));
        this.parent = null;
    }

    public void add(BookmarkItem item) {
        list.add(item);
    }

    public void add(BookmarkFolder folder) {
        list.add(folder);
    }

    public void addFirst(BookmarkFolder folder) {
        list.add(0, folder);
    }

    public BookmarkItem get(int index) {
        return list.get(index);
    }

    public List<BookmarkItem> getItemList() {
        return list;
    }

    public int size() {
        return list.size();
    }

    public void clear() {
        list.clear();
    }

    @Override
    protected int getType() {
        return BOOKMARK_ITEM_ID;
    }

    @Override
    protected boolean writeMain(JsonGenerator generator) throws IOException {
        generator.writeArrayFieldStart(COLUMN_NAME_LIST);
        for (BookmarkItem item : list) {
            if (!item.write(generator)) return false;
        }
        generator.writeEndArray();
        return true;
    }

    @Override
    protected boolean readMain(JsonParser parser) throws IOException {
        if (!COLUMN_NAME_LIST.equals(parser.getCurrentName())) return false;

        if (parser.nextToken() != JsonToken.START_ARRAY) return false;
        while (parser.nextToken() != JsonToken.END_ARRAY) {
            BookmarkItem item = read(parser, this);
            if (item == null) return false;
            list.add(item);
        }
        return true;
    }

    public final boolean writeForRoot(JsonGenerator generator) throws IOException {
        generator.writeStartArray();
        for (BookmarkItem item : list) {
            if (!item.write(generator)) return false;
        }
        generator.writeEndArray();
        return true;
    }

    public final boolean readForRoot(JsonParser parser) throws IOException {
        if (parser.nextToken() != JsonToken.START_ARRAY) return false;
        while (parser.nextToken() != JsonToken.END_ARRAY) {
            BookmarkItem item = read(parser, this);
            if (item == null) return false;
            list.add(item);
        }
        return true;
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/bookmark/BookmarkFolder.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.bookmark

import android.text.TextUtils
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.legacy.bookmark.util.BookmarkIdGenerator
import jp.hazuki.yuzubrowser.legacy.tab.manager.MainTabData
import java.io.IOException
import java.io.Serializable
import java.util.*

class BookmarkFolder : BookmarkItem, Serializable {

    var parent: BookmarkFolder? = null
    internal val list: ArrayList<BookmarkItem>

    val itemList: List<BookmarkItem>
        get() = list

    override val type: Int
        get() = BOOKMARK_ITEM_ID

    constructor(title: String?, parent: BookmarkFolder?, id: Long) : super(title, id) {
        this.list = ArrayList()
        this.parent = parent
    }

    constructor(list: List<MainTabData>, id: Long) : super(null, id) {
        this.list = ArrayList(list.size)
        for (tab in list)
            if (!TextUtils.isEmpty(tab.url))
                this.list.add(BookmarkSite(if (tab.title != null) tab.title else tab.url, tab.url, BookmarkIdGenerator.getNewId()))
        this.parent = null
    }

    fun add(item: BookmarkItem) {
        list.add(item)
    }

    fun add(folder: BookmarkFolder) {
        list.add(folder)
    }

    fun addFirst(folder: BookmarkFolder) {
        list.add(0, folder)
    }

    operator fun get(index: Int): BookmarkItem {
        return list[index]
    }

    fun size(): Int {
        return list.size
    }

    fun clear() {
        list.clear()
    }

    @Throws(IOException::class)
    override fun writeMain(writer: JsonWriter): Boolean {
        writer.name(COLUMN_NAME_LIST)
        writer.beginArray()
        list.forEach {
            if (!it.write(writer)) return false
        }
        writer.endArray()
        return true
    }

    @Throws(IOException::class)
    override fun readMain(name: String?, reader: JsonReader): Boolean {
        if (COLUMN_NAME_LIST != name) return false
        if (reader.peek() != JsonReader.Token.BEGIN_ARRAY) return false
        reader.beginArray()
        while (reader.hasNext()) {
            list.add(read(reader, this) ?: return false)
        }
        reader.endArray()
        return true
    }

    @Throws(IOException::class)
    fun writeForRoot(writer: JsonWriter): Boolean {
        writer.beginArray()
        list.forEach {
            if (!it.write(writer)) return false
        }
        writer.endArray()
        return true
    }

    @Throws(IOException::class)
    fun readForRoot(reader: JsonReader): Boolean {
        if (reader.peek() != JsonReader.Token.BEGIN_ARRAY) return false
        reader.beginArray()
        while (reader.hasNext()) {
            list.add(read(reader, this) ?: return false)
        }
        reader.endArray()
        return true
    }

    companion object {
        protected const val COLUMN_NAME_LIST = "2"
        const val BOOKMARK_ITEM_ID = 1
    }
}
